### ISSUE TITLE
DNS: Problem resolving services between pods.

### ISSUE THREAD

**Issue by Tiboris** [openshift/origin#13401]
Hi, 

when following tutorial to create statefulset with headless service nginx:
https://kubernetes.io/docs/tutorials/stateful-application/basic-stateful-set/ 
with slightly edited web.yaml as shown:
``` yaml
---
apiVersion: v1
kind: Service
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  ports:
  - port: 8080
    name: web
  clusterIP: None
  selector:
    app: nginx
---
apiVersion: apps/v1beta1
kind: StatefulSet
metadata:
  name: web
spec:
  serviceName: "nginx"
  replicas: 3
  template:
    metadata:
      labels:
        app: nginx
    spec:
      terminationGracePeriodSeconds: 10
      containers:
      - name: nginx
        image: gcr.io/google_containers/nginx-slim:0.8
        ports:
        - containerPort: 8080
          name: web
```
I bumped into problem with resolving other pods.

##### Version
```
# oc version
oc v1.4.1
kubernetes v1.4.0+776c994
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://192.168.122.254:8443
openshift v1.5.0-alpha.3+cf7e336
kubernetes v1.5.2+43a9be4
```

##### Steps To Reproduce
1. `oc cluster up --host-data-dir /var/lib/origin-data --use-existing-config --public-hostname=$(hostname) --version=v1.5.0-alpha.3`
2. `oc new project web`
3. `oc create -f web.yaml`

##### Current Result
```
# oc get pods
NAME      READY     STATUS    RESTARTS   AGE
web-0     1/1       Running   0          1m
web-1     1/1       Running   0          1m
web-2     1/1       Running   0          1m
```
```
# oc rsh  web-0 bash
bash-4.3$ echo $HOSTNAME
web-0
bash-4.3$ curl web-0
curl: (7) Failed to connect to web-0 port 80: Connection refused
bash-4.3$ curl web-1
curl: (6) Could not resolve host: web-1
bash-4.3$ cat /etc/resolv.conf 
search web.svc.cluster.local svc.cluster.local cluster.local
nameserver 192.168.122.254
options ndots:5
```
##### Expected Result

Resolved host web-1.

**Comment by mfojtik** [comment]
@csrwng @bparees I tested this on latest master without `oc cluster up` and DNS resolution inside the container worked well. However, in `oc cluster up` I cannot resolve the pod names from the stateful set, wonder if this is not cluster up resolver issue (but seems strange).

**Comment by Tiboris** [comment]
I have just tried same setup with v1.5.0-rc.0 origin image. Same result.
```
# oc run -i --tty --image busybox dns-test --restart=Never --rm /bin/sh      
Waiting for pod test/dns-test to be running, status is Pending, pod ready: false
Waiting for pod test/dns-test to be running, status is Pending, pod ready: false
If you don't see a command prompt, try pressing enter.
/ # nslookup web-0.nginx
Server:    192.168.122.254
Address 1: 192.168.122.254

nslookup: can't resolve 'web-0.nginx'
/ # nslookup web-1.nginx
Server:    192.168.122.254
Address 1: 192.168.122.254

nslookup: can't resolve 'web-1.nginx'
```

**Comment by bparees** [comment]
@csrwng any known fixes in this space?  possibly https://github.com/openshift/origin/pull/13539 ?

**Comment by csrwng** [comment]
Not specifically. I'll try with the latest master and see if it's still an issue.

**Comment by mukundpadale** [comment]
Is there any fix for this ? I am getting same issue while creating Cassandra cluster.
Stateful sets pods are not able to contact each other through DNS names , however they are able to ping using IP address.

**Comment by csrwng** [comment]
@mukundpadale what does your environment look like? Are you using cluster up to bring up your cluster? and can you include the output of 'oc version' ?

**Comment by mukundpadale** [comment]
I have deployed cassandra with statfulset. Creating 2 node cluster for cassandra we have to update seeds address into configuration file(cassandra.yaml) of each cassandra node.
but looks like statefulset1 can resolve its own DNS name entry not others i.e. statefulset2 and so on.

output of oc version,
oc v3.6.0+c4dd4cf
kubernetes v1.6.1+5115d708d7

Server https://<xyz>:443
openshift v1.5.1
kubernetes v1.5.2+43a9be4

I have used following yaml file for creating statefulset in openshift.
- kind: Service
  apiVersion: v1
  metadata:
    name: cassandra
    labels:
      name: cassandra
  spec:
    clusterIP: 'None' 
    ports:
    - port: 9042
    selector:
      name: cassandra

- kind: StatefulSet
  apiVersion: apps/v1beta1
  metadata:
    name: cassandra
  spec:
    serviceName: cassandra
    replicas: 2
    template:
      metadata:
        labels:
          app: cassandra
      spec:
        containers:
        - name: cassandra
          image: registry.acp.us-east-1.aws.symcpe.net:443/optimus/cassandra:1.0.0 
          imagePullPolicy: Always
          ports:
          - containerPort: 7000
            name: intra-node
          - containerPort: 7001
            name: tls-intra-node
          - containerPort: 7199
            name: jmx
          - containerPort: 9042
            name: cql
          resources:
            limits:
              cpu: "500m"
              memory: 5Gi
            requests:
             cpu: "500m"
             memory: 5Gi
          lifecycle:
            preStop:
              exec:
                command: ["/bin/sh", "-c", "PID=$(pidof java) && kill $PID && while ps -p $PID > /dev/null; do sleep 1; done"]
          env:
            - name: MAX_HEAP_SIZE
              value: 512M
            - name: HEAP_NEWSIZE
              value: 100M
            - name: CASSANDRA_SEEDS
              value: "cassandra-0.cassandra.cassandra.svc.cluster.local,cassandra-1.cassandra.cassandra.svc.cluster.local"
            - name: POD_IP
              valueFrom:
                fieldRef:
                  fieldPath: status.podIP
          readinessProbe:
            exec:
              command:
              - /bin/bash
              - -c
              - /ready-probe.sh
            initialDelaySeconds: 15
            timeoutSeconds: 5
        # These volume mounts are persistent. They are like inline claims,
        # but not exactly because the names need to match exactly one of
        # the stateful pod volumes.
          volumeMounts:
          - name: cassandra-data
            mountPath: /cassandra_data
  # These are converted to volume claims by the controller
  # and mounted at the paths mentioned above.
  # do not use these in production until ssd GCEPersistentDisk or other ssd pd
    volumeClaimTemplates:
    - metadata:
        name: cassandra-data
        annotations:
          volume.beta.kubernetes.io/storage-class: fast
      spec:
        accessModes: [ "ReadWriteOnce" ]
        resources:
          requests:
            storage: 1Gi

Here, namespace is - cassandra
headless service name - cassandra
statefulset name called cassandra has two pods - cassandra-0 and cassandra-1
am i missing anything? 

I have used this reference link for implementing cassandra 
https://kubernetes.io/docs/tutorials/stateful-application/cassandra

**Comment by LarryWax** [comment]
Hi, I have the same issue with DNS and stateful sets, pods that belongs to the same stateful set cannot resolv their hostname or hostnames from the other pods of the set.

The problem is that pods FQDN inside stateful set is in the format  $(podname).$(set service name).$(namespace).svc.cluster.local and not $(podname).$(namespace).svc.cluster.local like any other pods. [Stated here](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/)

The generated pod's resolv.conf doesn't contain $(set service name).$(namespace).svc.cluster.local as a search domain so is impossible for the pod to resolv hostnames that belongs to the stateful set.

My workaround was obviously to add $(set service name).$(namespace).svc.cluster.local to the search domains.

**Comment by simonpasquier** [comment]
I'm hitting this issue as well using oc cluster up (v3.7).

### How to reproduce

Apply this template

```
apiVersion: template.openshift.io/v1
kind: Template

metadata:
  name: example

objects:
# Headless service to resolve pod names
- apiVersion: v1
  kind: Service
  metadata:
    name: example
  spec:
    ports:
    - name: example
      port: 8080
      protocol: TCP
      targetPort: 8080
    clusterIP: None
    selector:
      app: example

- apiVersion: apps/v1beta1
  kind: StatefulSet
  metadata:
    name: example
  spec:
    selector:
      matchLabels:
        app: example
    updateStrategy:
      type: RollingUpdate
    podManagementPolicy: Parallel
    replicas: 2
    template:
      metadata:
        labels:
          app: example
        name: example
      spec:
        containers:
        - name: example
          args:
          - /bin/sleep
          - "3600"
          image: debian:stretch
          imagePullPolicy: IfNotPresent
        restartPolicy: Always
```

After deployment:

```
$ oc get all
NAME           READY     STATUS    RESTARTS   AGE
po/example-0   1/1       Running   0          1m
po/example-1   1/1       Running   0          1m

NAME          CLUSTER-IP   EXTERNAL-IP   PORT(S)    AGE
svc/example   None         <none>        8080/TCP   1m

NAME                   DESIRED   CURRENT   AGE
statefulsets/example   2         2         1m
```

Then from one of the containers, try to ping the other one using its name:

```
$ oc exec -ti example-0 ping example-1.example
ping: unknown host
command terminated with exit code 1
```

### Analysis

AFAICT this is because the Endpoints resource doesn't have the `hostname` fields defined which is what the DNS resolver looks up to resolve `example-0.example.<namespace>.svc.cluster.local.`. See [here](https://github.com/openshift/origin/blob/61e6f2584193fb08a812cd63b89dd4a09048f4dd/pkg/dns/serviceresolver.go#L224) and [here](https://github.com/openshift/origin/blob/61e6f2584193fb08a812cd63b89dd4a09048f4dd/pkg/dns/serviceresolver.go#L406-L412).

When I edit the endpoints definition to add the proper hostnames, it works (see below).

#### Before

```
apiVersion: v1
kind: Endpoints
metadata:
  creationTimestamp: 2017-12-13T14:45:23Z
  name: example
  namespace: myproject
  resourceVersion: "35182"
  selfLink: /api/v1/namespaces/myproject/endpoints/example
  uid: 3fa9e245-e014-11e7-bce4-0800274bb891
subsets:
- addresses:
  - ip: 192.168.69.3
    nodeName: localhost
    targetRef:
      kind: Pod
      name: example-0
      namespace: myproject
      resourceVersion: "35175"
      uid: 4068dabf-e014-11e7-bce4-0800274bb891
  - ip: 192.168.69.4
    nodeName: localhost
    targetRef:
      kind: Pod
      name: example-1
      namespace: myproject
      resourceVersion: "35179"
      uid: 4151d645-e014-11e7-bce4-0800274bb891
  ports:
  - name: example
    port: 8080
    protocol: TCP
```

#### After

```
apiVersion: v1
kind: Endpoints
metadata:
  creationTimestamp: 2017-12-13T14:45:23Z
  name: example
  namespace: myproject
  resourceVersion: "35394"
  selfLink: /api/v1/namespaces/myproject/endpoints/example
  uid: 3fa9e245-e014-11e7-bce4-0800274bb891
subsets:
- addresses:
  - hostname: example-0
    ip: 192.168.69.3
    nodeName: localhost
    targetRef:
      kind: Pod
      name: example-0
      namespace: myproject
      resourceVersion: "35175"
      uid: 4068dabf-e014-11e7-bce4-0800274bb891
  - hostname: example-1
    ip: 192.168.69.4
    nodeName: localhost
    targetRef:
      kind: Pod
      name: example-1
      namespace: myproject
      resourceVersion: "35179"
      uid: 4151d645-e014-11e7-bce4-0800274bb891
  ports:
  - name: example
    port: 8080
    protocol: TCP
```

```
$ oc exec -ti example-0 -- ping -c 1 example-1.example
PING example-1.example.myproject.svc.cluster.local (192.168.69.4): 56 data bytes
64 bytes from 192.168.69.4: icmp_seq=0 ttl=64 time=0.271 ms
--- example-1.example.myproject.svc.cluster.local ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max/stddev = 0.271/0.271/0.271/0.000 ms
```

### Version

```
oc v3.7.0+7ed6862
kubernetes v1.7.6+a08f5eeb62
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://10.0.2.15:8443
openshift v3.7.0+7ed6862
kubernetes v1.7.6+a08f5eeb62
```

**Comment by bparees** [comment]
I can't recreate this:

```
$ oc cluster up
Starting OpenShift using openshift/origin:v3.7.0 ...
Pulling image openshift/origin:v3.7.0
Pulled 2/4 layers, 54% complete
Pulled 3/4 layers, 93% complete
Pulled 4/4 layers, 100% complete
Extracting
Image pull complete
OpenShift server started.

The server is accessible via web console at:
    https://127.0.0.1:8443

You are logged in as:
    User:     developer
    Password: <any value>

To login as administrator:
    oc login -u system:admin
```
```

origin ((v3.7.0))$ oc new-app -f examples/sample-app/application-template-stibuild.json 
--> Deploying template "myproject/ruby-helloworld-sample" for "examples/sample-app/application-template-stibuild.json" to project myproject

     ruby-helloworld-sample
     ---------
     This example shows how to create a simple ruby application in openshift origin v3

     * With parameters:
        * MYSQL_USER=userTB8 # generated
        * MYSQL_PASSWORD=KEv0jvb3 # generated
        * MYSQL_DATABASE=root

--> Creating resources ...
    secret "dbsecret" created
    service "frontend" created
    route "route-edge" created
    imagestream "origin-ruby-sample" created
    imagestream "ruby-22-centos7" created
    buildconfig "ruby-sample-build" created
    deploymentconfig "frontend" created
    service "database" created
    deploymentconfig "database" created
--> Success
    Access your application via route 'www.example.com' 
    Build scheduled, use 'oc logs -f bc/ruby-sample-build' to track its progress.
    Run 'oc status' to view your app.
```

```
origin ((v3.7.0))$ oc get pods
NAME                        READY     STATUS      RESTARTS   AGE
database-1-p9d7r            1/1       Running     0          2m
frontend-1-fqf92            1/1       Running     0          1m
frontend-1-jsbjc            1/1       Running     0          1m
ruby-sample-build-1-build   0/1       Completed   0          2m
```
```
origin ((v3.7.0))$ oc rsh frontend-1-jsbjc
sh-4.2$ ping database
PING database.myproject.svc.cluster.local (172.30.215.238) 56(84) bytes of data.
^C
--- database.myproject.svc.cluster.local ping statistics ---
1 packets transmitted, 0 received, 100% packet loss, time 0ms

sh-4.2$ ping frontend
PING frontend.myproject.svc.cluster.local (172.30.105.17) 56(84) bytes of data.
^C
--- frontend.myproject.svc.cluster.local ping statistics ---
1 packets transmitted, 0 received, 100% packet loss, time 0ms
```

**Comment by bparees** [comment]
and my endpoints (no hostnames indicated):
```
$ oc get endpoints -o yaml
apiVersion: v1
items:
- apiVersion: v1
  kind: Endpoints
  metadata:
    creationTimestamp: 2017-12-13T20:37:27Z
    labels:
      app: ruby-helloworld-sample
      template: application-template-stibuild
    name: database
    namespace: myproject
    resourceVersion: "2645"
    selfLink: /api/v1/namespaces/myproject/endpoints/database
    uid: 6eabea64-e045-11e7-85c6-507b9d27b5d9
  subsets:
  - addresses:
    - ip: 172.17.0.4
      nodeName: localhost
      targetRef:
        kind: Pod
        name: database-1-p9d7r
        namespace: myproject
        resourceVersion: "2642"
        uid: 7c67b156-e045-11e7-85c6-507b9d27b5d9
    ports:
    - name: db
      port: 3306
      protocol: TCP
- apiVersion: v1
  kind: Endpoints
  metadata:
    creationTimestamp: 2017-12-13T20:37:27Z
    labels:
      app: ruby-helloworld-sample
      template: application-template-stibuild
    name: frontend
    namespace: myproject
    resourceVersion: "2776"
    selfLink: /api/v1/namespaces/myproject/endpoints/frontend
    uid: 6e93cbbf-e045-11e7-85c6-507b9d27b5d9
  subsets:
  - addresses:
    - ip: 172.17.0.3
      nodeName: localhost
      targetRef:
        kind: Pod
        name: frontend-1-jsbjc
        namespace: myproject
        resourceVersion: "2766"
        uid: 91d2813c-e045-11e7-85c6-507b9d27b5d9
    - ip: 172.17.0.6
      nodeName: localhost
      targetRef:
        kind: Pod
        name: frontend-1-fqf92
        namespace: myproject
        resourceVersion: "2775"
        uid: 91c57077-e045-11e7-85c6-507b9d27b5d9
    ports:
    - name: web
      port: 8080
      protocol: TCP
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

**Comment by simonpasquier** [comment]
@bparees IIUC you aren't testing exactly the same thing. You've checked the service's DNS name which works fine for me too (same for the OP I guess).

What isn't working is the DNS resolution for the individual endpoint names of the **StatefulSet** resource (eg example-0.example in my setup). To be fair, the [OpenShift documentation](https://docs.openshift.org/latest/architecture/networking/networking.html#architecture-additional-concepts-openshift-dns) doesn't mention this capability while the [Kubernetes documentation](https://v1-7.docs.kubernetes.io/docs/tutorials/stateful-application/basic-stateful-set/#using-stable-network-identities) does.

**Comment by bparees** [comment]
> What isn't working is the DNS resolution for the individual endpoint names of the StatefulSet resource (eg example-0.example in my setup). To be fair, the OpenShift documentation doesn't mention this capability while the Kubernetes documentation does.

Ok, thanks for clarifying.  So this does not sound like a cluster up specific DNS issue, it is a question about kubernetes statefulset hostname resolution behavior.  Reassigning.

**Comment by smarterclayton** [comment]
In the most recent example, you need to set the service reference from the statefulset in order for the hostname field on endpoints to be populated. Did you do that?

**Comment by simonpasquier** [comment]
@smarterclayton my bad, i didn't do it but setting `serviceName` in the statefulset fixed my issue indeed.

**Comment by simonpasquier** [comment]
I've also retried the example from @Tiboris (the OP) and I can't reproduce the issue...

**Comment by smarterclayton** [comment]
Going to mark as closed (significant fixes to DNS for statefulsets were part of 1.6), please reopen if you're still seeing issues.

**Comment by tech-mint** [comment]
still having the issue on OKD 3.9. It seems there is a related work around in here:
https://github.com/openshift/origin/issues/21354
